File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    // two-pass solution, O(n) time, O(1) space
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length = 0;
        ListNode* curr = head;
        while (curr) {
            length++;
            curr = curr->next;
        }

        int removeIndex = length - n;
        // if the node needs to be removed is the head
        if (removeIndex == 0) {
            ListNode* temp = head;
            head = head->next;
            delete temp;
            return head;
        }

        curr = head;
        for (int i = 0; i < length - 1; i++) {
            if (i == removeIndex - 1) {
                ListNode* temp = curr->next;
                curr->next = curr->next->next;
                delete temp;
                break;
            }
            curr = curr->next;
        }
        return head;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    # reverse and then merge solution, O(n) time, O(1) space
    # pass 1: find the middle of the list
    # pass 2: reverse the second half of the list
    # pass 3: merge the first half and the reversed second half
    def reorderList(self, head: Optional[ListNode]) -> None:
        # pass 1: find the middle part of the list (the slow pointer)
        # once fast reaches the end, slow will point to the middle
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # pass 2: reverse the second half of the list
        middle_curr = slow.next # starting point of the half
        slow.next = None # disconnect the two halves before the merge
        prev = None # this will be the head of the reversed half
        while middle_curr:
            next = middle_curr.next
            middle_curr.next = prev
            prev = middle_curr
            middle_curr = next

        # pass 3: merge the two halves
        first_curr, second_curr = head, prev
        while second_curr:
            first_next, second_next = first_curr.next, second_curr.next
            first_curr.next = second_curr
            second_curr.next = first_next
            first_curr, second_curr = first_next, second_next
        